### PR TITLE
ENS L2

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/22930
Sec review: https://github.com/brave/security/issues/966

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Behind a feature flag `BraveWalletENSL2`
With this PR wallet panel 'Send to' input should resolve `offchainexample.eth` to `0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045`

### CODE REVIEW

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (diracdeltas) [inline components/brave_wallet/browser/json_rpc_service.cc:96]: is there not something in Chromium we can use to do this instead of writing it ourselves? cc @bridiver

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (diracdeltas) [inline components/brave_wallet/browser/ens_resolver_task.cc:233]: is something passing the JSON from the network through the JSON sanitizer before `internal_callback` gets called?

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc:96]: There is similar thing in chrome https://source.chromium.org/chromium/chromium/src/+/main:net/dns/dns_util.cc;l=47?q=DNSDomainFromDot&sq=&ss=chromium
But per ENS spec https://docs.ens.domains/ens-improvement-proposals/ensip-10-wildcard-resolution#specification there is no total length limitation, so I needed a custom converter.

Will try to check if >256 char length domains actually exist in ENS, if not maybe it makes sense to use chrome's implementation.

Still needs a better comment here.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc:233]: Good catch, it's missing here. Will fix.
Thanks!

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (thypon) [inline components/brave_wallet/browser/json_rpc_service.cc:106]: this math assigning a possibly signed to unsigned seems would be better with an extra check.
Did we tested for UTF-8 strings anyway?

**Review by thypon — COMMENTED** [review]
Are we validating for strings containing only allowed characters and non-UTF8 strings/chars here?

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc:186]: OnEnsResolverTaskDone method deletes task object which is not obvious from its name, so such comment takes place there. Common pattern through chromium code.

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc:106]: This math is all unsigned
Will add [similar](https://source.chromium.org/chromium/chromium/src/+/main:net/dns/dns_util.cc;drc=bead2b4d2553e57229747b48b3dceaadbb8972bf;l=142?q=DNSDomainFromDot&ss=chromium) byte validation. Thanks!

Comment (thypon) [reply components/brave_wallet/browser/json_rpc_service.cc:106]: Not sure if it is the correct validation over there. https://docs.ens.domains/frequently-asked-questions `nameprep` seems the standard to follow here, to validate these names. Regarding these, we should be careful in considering multibyte runes when iterating through characters instead of single-byte sequences.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc:233]: Actually `json_rpc_service_base_->RequestInternal` will sanitize json internally, so nothing to add here.

**Review by Dzezo12345 — COMMENTED** [review]
- [ ] # - - @~~_[]()****_~~

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc:106]: Discussed in slack: we already support only ascii domain names for ENS, so no nameprep here. Will add explicit byte validation to DnsEncode.

Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc]: https://github.com/brave/brave-core/pull/14447/commits/b398bac78f1c7e868d6e49ab6aa9fc0be5ac4c37

File: components/brave_wallet/browser/eth_data_builder_unittest.cc
Comment (thypon) [inline components/brave_wallet/browser/eth_data_builder_unittest.cc:323]: Are `IP` alike valid DnsEncoded ENS?

**Review by thypon — COMMENTED** [review]
⬆️

File: components/brave_wallet/browser/eth_data_builder_unittest.cc
Comment (supermassive) [reply components/brave_wallet/browser/eth_data_builder_unittest.cc:323]: Technically DnsEncode of an ipaddress will result in something meaningful. I don't any restriction why it shouldn't.
In our case all ENS stuff happens for .eth tld.
In [reference implementation](https://github.com/ensdomains/offchain-resolver/blob/main/packages/gateway/src/server.ts#L30-L40 ) gateway is interpreting that encoded name to extract required record. Makes most sense for wildcard case which is not part of this PR

Actually I'm not sure I understand why ENS designers brought DNSEncoding here - plain string would be enough. Now it looks to me as unjustified complication.

**Review by thypon — APPROVED** [review]
Code seems to be good from a security perspective. Wait for the S&P review to close before merging!

File: components/brave_wallet/browser/eip1559_transaction.cc
Comment (supermassive) [inline components/brave_wallet/browser/eip1559_transaction.cc:259]: made EthAddress::bytes() to return const &, so some places needed a movable clone.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/ens_resolver_task.cc:454]: I see `JsonReader::Read` will be called by `ExtractGatewayResult` in `OnFetchOffchainDone`. Isn't this response body unsanitized if we use url loader directly instead of our api_request_helper? Also, what's the reason to use url loader directly rather than using api_request_helper?

File: components/brave_wallet/browser/json_rpc_service_base.h
Comment (yrliou) [inline components/brave_wallet/browser/json_rpc_service_base.h:18]: Could you explain what was the reason adding this base class?

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/ens_resolver_task.cc:76]: This function is another reason that I would prefer we utilize api request helper if possible.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (yrliou) [inline components/brave_wallet/browser/ens_resolver_task.cc:113]: records_v->GetDict().FindString

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc:40]: Well this is a json external to chromium, so it should not have comments etc.

File: components/brave_wallet/browser/eth_data_builder.cc
Comment (supermassive) [reply components/brave_wallet/browser/eth_data_builder.cc]: Made it to use TupleEncoder

File: components/brave_wallet/browser/eth_data_builder.cc
Comment (supermassive) [reply components/brave_wallet/browser/eth_data_builder.cc]: No need in this function actually. Removed

File: components/brave_wallet/browser/json_rpc_service_base.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service_base.cc]: done.

File: components/brave_wallet/browser/json_rpc_service_base.h
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service_base.h]: done.

File: components/brave_wallet/browser/json_rpc_response_parser.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_response_parser.cc:15]: done.

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc]: removed todo

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc]: done.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc]: done.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc]: Made it so ENSResolver uses ApiRequestHelper now

File: components/brave_wallet/browser/json_rpc_service_base.h
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service_base.h]: removed base class

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc]: done.

File: components/brave_wallet/browser/ens_resolver_task.cc
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.cc]: done.

**Review by yrliou — COMMENTED** [review]
Overall looks good just a few comment left.

File: components/brave_wallet/common/eth_abi_utils_unittest.cc
Comment (supermassive) [reply components/brave_wallet/common/eth_abi_utils_unittest.cc:67]: done

File: components/brave_wallet/browser/ens_resolver_task.h
Comment (supermassive) [reply components/brave_wallet/browser/ens_resolver_task.h:70]: done

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service.cc]: done

**Review by yrliou — APPROVED** [review]
++ assuming CI is green.

**Pull request merged**
